<a href="https://colab.research.google.com/github/Mix1996/Project-2-Classification/blob/main/Project_2_Part_1_(Core).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### **Import Libraries**

In [17]:
## Pandas
import pandas as pd
## Numpy
import numpy as np


#**Load the Data Dataset 1**

In [18]:
## Load in Data
path = 'https://docs.google.com/spreadsheets/d/e/2PACX-1vTcEGRHb41Faze2NRGsr7_flDwSaTRzfYVkzpuFmBDmW6JYaMD7HKt7ahPJAhaPpxuH89scIHSuTm_O/pub?output=csv'
df = pd.read_csv(path)

### **Inspect the Data**

In [19]:
## Display the first (5) rows
df.head()

age  workclass  fnlwgt     education  educational-num      marital-status  \
0   25    Private  226802          11th                7       Never-married   
1   38    Private   89814       HS-grad                9  Married-civ-spouse   
2   28  Local-gov  336951    Assoc-acdm               12  Married-civ-spouse   
3   44    Private  160323  Some-college               10  Married-civ-spouse   
4   18          ?  103497  Some-college               10       Never-married   

          occupation relationship   race  gender  capital-gain  capital-loss  \
0  Machine-op-inspct    Own-child  Black    Male             0             0   
1    Farming-fishing      Husband  White    Male             0             0   
2    Protective-serv      Husband  White    Male             0             0   
3  Machine-op-inspct      Husband  Black    Male          7688             0   
4                  ?    Own-child  White  Female             0             0   

   hours-per-week native-country income  
0              40  United-States  <=50K  
1              50  United-States  <=50K  
2              40  United-States   >50K  
3              40  United-States   >50K  
4              30  United-States  <=50K

In [20]:
## Display the number of rows and columns for the dataframe
df.shape
print(f'There are {df.shape[0]} rows, and {df.shape[1]} columns.')
print(f'The rows represent {df.shape[0]} observations, and the columns represent {df.shape[1]-1} features and 1 target variable.')

There are 48842 rows, and 15 columns.
The rows represent 48842 observations, and the columns represent 14 features and 1 target variable.


In [21]:
## Display the column names, count of non-null values, and their datatypes
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48842 entries, 0 to 48841
Data columns (total 15 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   age              48842 non-null  int64 
 1   workclass        48842 non-null  object
 2   fnlwgt           48842 non-null  int64 
 3   education        48842 non-null  object
 4   educational-num  48842 non-null  int64 
 5   marital-status   48842 non-null  object
 6   occupation       48842 non-null  object
 7   relationship     48842 non-null  object
 8   race             48842 non-null  object
 9   gender           48842 non-null  object
 10  capital-gain     48842 non-null  int64 
 11  capital-loss     48842 non-null  int64 
 12  hours-per-week   48842 non-null  int64 
 13  native-country   48842 non-null  object
 14  income           48842 non-null  object
dtypes: int64(6), object(9)
memory usage: 5.6+ MB


In [22]:
## Display the descriptive statistics for the numeric columns
df.describe(include="number") # or 'object'

age        fnlwgt  educational-num  capital-gain  \
count  48842.000000  4.884200e+04     48842.000000  48842.000000   
mean      38.643585  1.896641e+05        10.078089   1079.067626   
std       13.710510  1.056040e+05         2.570973   7452.019058   
min       17.000000  1.228500e+04         1.000000      0.000000   
25%       28.000000  1.175505e+05         9.000000      0.000000   
50%       37.000000  1.781445e+05        10.000000      0.000000   
75%       48.000000  2.376420e+05        12.000000      0.000000   
max       90.000000  1.490400e+06        16.000000  99999.000000   

       capital-loss  hours-per-week  
count  48842.000000    48842.000000  
mean      87.502314       40.422382  
std      403.004552       12.391444  
min        0.000000        1.000000  
25%        0.000000       40.000000  
50%        0.000000       40.000000  
75%        0.000000       45.000000  
max     4356.000000       99.000000

## **First choice: dataset 1**

**1.Source of data**
   This dataset named “adult” is found in the UCI machine learning repository
   http://www.cs.toronto.edu/~delve/data/adult/desc.html
   
   The detailed description on the dataset can be found in the original UCI documentation
   http://www.cs.toronto.edu/~delve/data/adult/adultDetail.html

**2. Brief description of data**

This dataset appears to be related to income classification or prediction, where the goal might be to predict an individual's income category (e.g., ">50K" or "<=50K") based on various features such as age, education, occupation, etc.

Total Columns: 15
Total Rows: 48842

**3. What is the target?**

The target is the income. We can try predict the income range of an individual based on the feaures.

**4. What does one row represent? (A person? A business? An event? A product?)**

One row in this dataset represents information about a person. Each row provides demographic and employment-related attributes for an individual. Therefore, the dataset is focused on describing and profiling people, not businesses, events, or products.

**5. Is this a classification or regression problem?**

If the primary objective is to predict or classify whether an individual's income falls into one of two categories, such as ">50K" or "<=50K," then it is a classification problem.

**6. How many features does the data have?**
14 features (income is excluded as this is the target)

**7. How many rows are in the dataset?**

Total Rows: 48842

**8. What, if any, challenges do you foresee in cleaning, exploring, or modeling this dataset?**

**Categorical Data Handling**: Many columns in the dataset are categorical, such as "workclass," "education," "marital-status," "occupation," "relationship," "race," "gender," and "native-country." Handling categorical data often requires encoding techniques like one-hot encoding or label encoding.

**Missing Data:** Although the dataset is described as having no missing values, in real-world scenarios, missing data is common. It's essential to handle missing data if it exists, either by imputation or by removing rows or columns with missing values.

**Model Fairness:** Ensuring that the model does not exhibit bias or unfairness, especially when making decisions about individuals' income.

**Feature Importance:** Determining which features are the most important for predicting income, as this can provide valuable insights.


# **Load the Data Dataset 2**

In [23]:
## Load in Data
path = 'https://docs.google.com/spreadsheets/d/e/2PACX-1vSn83v5-8JzCpC7oPmBrXwzkP7zkqSEojOOS8Bf5Z_sDC-z68i9cjPbLRRXBdEi45BDarFIMEj9qLDB/pub?output=csv'
df2 = pd.read_csv(path)

### **Inspect the Data**

In [24]:
## Display the first (5) rows
df2.head()

id  gender   age  hypertension  heart_disease ever_married  \
0   9046    Male  67.0             0              1          Yes   
1  51676  Female  61.0             0              0          Yes   
2  31112    Male  80.0             0              1          Yes   
3  60182  Female  49.0             0              0          Yes   
4   1665  Female  79.0             1              0          Yes   

       work_type Residence_type  avg_glucose_level   bmi   smoking_status  \
0        Private          Urban             228.69  36.6  formerly smoked   
1  Self-employed          Rural             202.21   NaN     never smoked   
2        Private          Rural             105.92  32.5     never smoked   
3        Private          Urban             171.23  34.4           smokes   
4  Self-employed          Rural             174.12  24.0     never smoked   

   stroke  
0       1  
1       1  
2       1  
3       1  
4       1

In [25]:
## Display the number of rows and columns for the dataframe
df2.shape
print(f'There are {df2.shape[0]} rows, and {df.shape[1]} columns.')
print(f'The rows represent {df2.shape[0]} observations, and the columns represent {df2.shape[1]-1} features and 1 target variable.')

There are 5110 rows, and 15 columns.
The rows represent 5110 observations, and the columns represent 11 features and 1 target variable.


In [26]:
## Display the column names, count of non-null values, and their datatypes
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5110 entries, 0 to 5109
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   id                 5110 non-null   int64  
 1   gender             5110 non-null   object 
 2   age                5110 non-null   float64
 3   hypertension       5110 non-null   int64  
 4   heart_disease      5110 non-null   int64  
 5   ever_married       5110 non-null   object 
 6   work_type          5110 non-null   object 
 7   Residence_type     5110 non-null   object 
 8   avg_glucose_level  5110 non-null   float64
 9   bmi                4909 non-null   float64
 10  smoking_status     5110 non-null   object 
 11  stroke             5110 non-null   int64  
dtypes: float64(3), int64(4), object(5)
memory usage: 479.2+ KB


In [27]:
## Display the descriptive statistics for the numeric columns
df2.describe(include="number") # or 'object'

id          age  hypertension  heart_disease  \
count   5110.000000  5110.000000   5110.000000    5110.000000   
mean   36517.829354    43.226614      0.097456       0.054012   
std    21161.721625    22.612647      0.296607       0.226063   
min       67.000000     0.080000      0.000000       0.000000   
25%    17741.250000    25.000000      0.000000       0.000000   
50%    36932.000000    45.000000      0.000000       0.000000   
75%    54682.000000    61.000000      0.000000       0.000000   
max    72940.000000    82.000000      1.000000       1.000000   

       avg_glucose_level          bmi       stroke  
count        5110.000000  4909.000000  5110.000000  
mean          106.147677    28.893237     0.048728  
std            45.283560     7.854067     0.215320  
min            55.120000    10.300000     0.000000  
25%            77.245000    23.500000     0.000000  
50%            91.885000    28.100000     0.000000  
75%           114.090000    33.100000     0.000000  
max           271.740000    97.600000     1.000000

## **Second choice: dataset 2**

**1.Source of data**
   This dataset named “Stroke Prediction Dataset” is found Kaggle learning repository
   https://www.kaggle.com/datasets/fedesoriano/stroke-prediction-dataset
  

**2. Brief description of data**

This dataset is used to predict whether a patient is likely to get stroke based on the input parameters like gender, age, various diseases, and smoking status. Each row in the data provides relavant information about the patient.

Total Columns: 15
Total Rows: 5110

**3. What is the target?**

The target is the stroke to predict if the patient is going to have a stroke.

**4. What does one row represent? (A person? A business? An event? A product?)**

One row in this dataset represents information about an individual person. Each row provides various attributes and characteristics of an individual, such as their age, gender, health conditions (hypertension and heart disease), marital status, occupation, residence type, health metrics (average glucose level and BMI), smoking status, and whether or not they had a stroke.

**5. Is this a classification or regression problem?**

In classification tasks, the objective is to categorize data points (in this case, individuals) into one of the predefined classes or categories, which aligns with the binary classification of stroke risk in this dataset.

**6. How many features does the data have?**
11 features and one target variable

**7. How many rows are in the dataset?**

Total Rows: 5110

**8. What, if any, challenges do you foresee in cleaning, exploring, or modeling this dataset?**

**Handling Missing Data:** The 'bmi' column contains missing values (null values). Dealing with missing data can be challenging, and you'll need to decide on an appropriate strategy, such as imputation or removal of rows with missing values.

**Imbalanced Target Variable: **Imbalanced datasets, where one class (e.g., individuals who had a stroke) is significantly less frequent than the other, can lead to model bias.

**Model Selection and Evaluation:** Choosing an appropriate machine learning algorithm and evaluation metrics can be challenging.

**Interpreting Results:** Understanding the factors that contribute to stroke risk and interpreting the model's predictions can be complex. This is especially important in healthcare applications, where interpretability and transparency are critical.
